sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EqualsArgumentTypeCheck.kt

```diff
@@ -0,0 +1,106 @@
+        val childNames = klass.collectDescendantsOfType<KtClass>().mapNotNull { it.name }
+        val parentNames = klass.superTypeListEntries.mapNotNull { it.typeReference!!.nameForReceiverLabel() }
+
+        function.collectDescendantsOfType<KtIsExpression> { parameter.name == (it.leftHandSide as? KtNameReferenceExpression)?.getReferencedName() }
+            .any {
+                // typeReference is always present
+                val name = it.typeReference!!.nameForReceiverLabel()
+                childNames.contains(name) || parentNames.contains(name)
+            }
+            .let { if (it) return }
+
+        function.collectDescendantsOfType<KtBinaryExpression> { it.operationToken == KtTokens.EQEQ || it.operationToken == KtTokens.EXCLEQ }
+            .any { binaryExpression -> isBinaryExpressionCorrect(binaryExpression, parameter, klass) }
+            .let { if (it) return }
+
+        function.collectDescendantsOfType<KtBinaryExpressionWithTypeRHS> { it.operationReference.getReferencedName() == "as?" }
+            .any { binaryExpression -> isBinaryExpressionWithTypeCorrect(binaryExpression, parameter, klass) }
+            .let { if (it) return }
+
+        ctx.reportIssue(function.nameIdentifier!!, "Add a type test to this method.")
+    }
+
+    private fun isBinaryExpressionWithTypeCorrect(
+        binaryExpression: KtBinaryExpressionWithTypeRHS,
+        parameter: KtParameter,
+        klass: KtClass,
+    ): Boolean {
+        val left = binaryExpression.left.collectDescendantsOfType<KtNameReferenceExpression>().map { it.getReferencedName() }
+        val right = binaryExpression.right!!.collectDescendantsOfType<KtNameReferenceExpression>().map { it.getReferencedName() }
+        return (left.contains(parameter.name) && isContainedInList(right, klass)) || (isContainedInList(left, klass) && right.contains(
+            parameter.name
+        ))
+    }
+
+    private fun isBinaryExpressionCorrect(binaryExpression: KtBinaryExpression, parameter: KtParameter, klass: KtClass): Boolean {
+        val left = binaryExpression.left!!.collectDescendantsOfType<KtNameReferenceExpression>()
+            .filterNot { (it.parent is KtDotQualifiedExpression || it.parent is KtSafeQualifiedExpression) && it.getReferencedName() == "javaClass" }
+            .map { it.getReferencedName() }
+        val right = binaryExpression.right!!.collectDescendantsOfType<KtNameReferenceExpression>()
+            .filterNot { (it.parent is KtDotQualifiedExpression || it.parent is KtSafeQualifiedExpression) && it.getReferencedName() == "javaClass" }
+            .map { it.getReferencedName() }
+        return (left.contains(parameter.name) && isContainedInList(right, klass)) || (isContainedInList(left, klass) && right.contains(
+            parameter.name
+        ))
+    }
+
+    private fun isContainedInList(list: List<String>, klass: KtClass) =
+        list.contains(klass.name) || list.contains(THIS_KEYWORD.value) || list.contains("javaClass")
```

Comment:
how about naming this method `isCheckedAgainstRightType` or sth like this?

It's up to you and style, you can use a bit more functional approach `listOf(klass.name, THIS_KEYWORD.value, "javaClass").any { list.contains(it) }`